tests/integrationv2/processes.py

```diff
@@ -0,0 +1,86 @@
+import os
+import time
+import threading
+import subprocess
+from common import Results, TimeoutException
+
+
+class ManagedProcess(threading.Thread):
+    """
+    A ManagedProcess is a thread that monitors a subprocess.
+    This class provides a single place to control process timeouts and cleanup.
+
+    The stdin/stdout/stderr and exist code a monitored and results
+    are made available to the caller.
+    """
+    def __init__(self, cmd_line, provider_set_ready_condition, timeout):
+        threading.Thread.__init__(self)
+        self.cmd_line = cmd_line
+        self.timeout = timeout
+        self.results_condition = threading.Condition()
+        self.ready_condition = threading.Condition()
+        self.results = None
+        self.process_ready = False
+        self.provider_set_ready_condition = provider_set_ready_condition
+
+    def run(self):
+        with self.results_condition:
+            try:
+                proc = subprocess.Popen(self.cmd_line, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
+            except Exception as ex:
+                self.results = Results(None, None, None, ex)
+                raise ex
+
+            self.provider_set_ready_condition()
+
+            # Result should be available to the whole scope
+            proc_results = None
+            try:
+                proc_results = proc.communicate(timeout=self.timeout)
```

Comment:
If a ManagedProcess communciates and then exits, how do we handle tests that need to interact with the connection? Not all tests can be accomplished entirely through CLI arguments. For example, in order to trigger KeyUpdate, we'll need to send a specific command to the openssl process. See "CONNECTED COMMANDS" here: https://www.openssl.org/docs/man1.0.2/man1/openssl-s_server.html

Reply:
Discussed offline. This is a bit of an unknown due to `communicate()` failing with a pipe'd stdin.

Reply:
Oh I didn't really answer the first part.

`communicate` waits until the process ends. Since each peer is running in its own process, they can communicate asynchronously. The caller provides some kind of input, which gets sent along.

If we need more fine grained behavior like pexpect, that would be a whole other paradigm to consider.